plugins/router/f5/plugin.go

```diff
@@ -0,0 +1,529 @@
+				err)
+			return nil, err
+		}
+		if n != len(cfg.PrivateKey) {
+			glog.Errorf("Wrong number of bytes written to tempfile for private key"+
+				" for F5 BIG-IP: expected %d bytes, wrote %d",
+				len(cfg.PrivateKey), n)
+			return nil, err
+		}
+		err = privkeyFile.Close()
+		if err != nil {
+			glog.Errorf("Error closing tempfile for private key for F5 BIG-IP: %v",
+				err)
+			return nil, err
+		}
+
+		privkeyFileName = privkeyFile.Name()
+	}
+
+	f5LTMCfg := f5LTMCfg{
+		host:         cfg.Host,
+		username:     cfg.Username,
+		password:     cfg.Password,
+		httpVserver:  cfg.HttpVserver,
+		httpsVserver: cfg.HttpsVserver,
+		privkey:      privkeyFileName,
+		insecure:     cfg.Insecure,
+	}
+	f5, err := newF5LTM(f5LTMCfg)
+	if err != nil {
+		return nil, err
+	}
+	return &F5Plugin{f5}, f5.Initialize()
+}
+
+// ensurePoolExists checks whether the named pool already exists in F5 BIG-IP
+// and creates it if it does not.
+func (p *F5Plugin) ensurePoolExists(poolname string) error {
+	poolExists, err := p.F5Client.PoolExists(poolname)
+	if err != nil {
+		glog.V(4).Infof("F5Client.PoolExists failed: %v", err)
+		return err
+	}
+
+	if !poolExists {
+		err = p.F5Client.CreatePool(poolname)
+		if err != nil {
+			glog.V(4).Infof("Error creating pool %s: %v", poolname, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// updatePool update the named pool (which must already exist in F5 BIG-IP) with
+// the given endpoints.
+func (p *F5Plugin) updatePool(poolname string, endpoints *kapi.Endpoints) error {
+	members, err := p.F5Client.GetPoolMembers(poolname)
+	if err != nil {
+		glog.V(4).Infof("F5Client.GetPoolMembers failed: %v", err)
+		return err
+	}
+
+	// We need to keep track of which endpoints already existed in F5 in order
+	// to delete any that no longer exist in the updated set of endpoints.
+	//
+	// It would be really nifty if F5 would just let us PUT the new list to
+	// the pool members resource instead of making a bunch of POST and DELETE
+	// requests, bu-u-ut... it doesn't.
+	needToDelete := map[string]bool{}
+	for member, _ := range members {
+		if members[member] {
+			needToDelete[member] = true
+		}
+	}
+
+	// Note that OpenShift issues many spurious notifications for updates when
+	// the endpoints set is actually the same, so we may ultimately end up
+	// adding and deleting 0 endpoints.
+
+	for _, subset := range endpoints.Subsets {
+		for _, addr := range subset.Addresses {
+			for _, port := range subset.Ports {
+				dest := fmt.Sprintf("%s:%d", addr.IP, port.Port)
+				exists := needToDelete[dest]
+				needToDelete[dest] = false
+				if exists {
+					glog.V(4).Infof("  Skipping %s because it already exists.", dest)
+				} else {
+					glog.V(4).Infof("  Adding %s...", dest)
+					err = p.F5Client.AddPoolMember(poolname, dest)
+					if err != nil {
+						glog.V(4).Infof("  Error adding endpoint %s to pool %s: %v",
+							dest, poolname, err)
+					}
+				}
+			}
+		}
+	}
+
+	for member, _ := range needToDelete {
```

Comment:
Aah, I see what I missed - the ```needToDelete[dest] = false```.  Yeah my understanding of the comments above was that we were deleting the existing members and then re-adding them back in. I see now that this is adding the new members in first and then deleting the ones that are missing.